components/raftstore-v2/src/operation/query/lease.rs

```diff
@@ -0,0 +1,344 @@
+            notify_stale_req(self.term(), ch);
+            poll_ctx.raft_metrics.propose.dropped_read_index.inc();
+            return false;
+        }
+
+        let mut read = ReadIndexRequest::with_command(id, req, ch, now);
+        read.addition_request = request.map(Box::new);
+        self.pending_reads_mut().push_back(read, is_leader);
+        debug!(
+            self.logger,
+            "request to get a read index";
+            "request_id" => ?id,
+            "is_leader" => is_leader,
+        );
+
+        true
+    }
+
+    fn read_index_leader<T: Transport>(
+        &mut self,
+        poll_ctx: &mut StoreContext<EK, ER, T>,
+        mut req: RaftCmdRequest,
+        ch: QueryResChannel,
+    ) {
+        let now = monotonic_raw_now();
+        let lease_state = self.inspect_lease();
+        if can_amend_read::<QueryResChannel>(
+            self.pending_reads().back(),
+            &req,
+            lease_state,
+            poll_ctx.cfg.raft_store_max_leader_lease(),
+            now,
+        ) {
+            // Must use the commit index of `PeerStorage` instead of the commit index
+            // in raft-rs which may be greater than the former one.
+            // For more details, see the annotations above `on_leader_commit_idx_changed`.
+            let commit_index = self.storage().entry_storage().commit_index();
+            if let Some(read) = self.pending_reads_mut().back_mut() {
+                // A read request proposed in the current lease is found; combine the new
+                // read request to that previous one, so that no proposing needed.
+                read.push_command(req, ch, commit_index);
+                return;
+            }
+        }
+
+        if self.propose_read_index(poll_ctx, req, self.is_leader(), ch, now) {
+            self.set_has_ready();
+        }
+
+        // TimeoutNow has been sent out, so we need to propose explicitly to
+        // update leader lease.
+        // TODO:add following when propose is done
+        // if self.leader_lease.is_suspect() {
+        // let req = RaftCmdRequest::default();
+        // if let Ok(Either::Left(index)) = self.propose_normal(poll_ctx, req) {
+        // let (callback, _) = CmdResChannel::pair();
+        // let p = Proposal {
+        // is_conf_change: false,
+        // index,
+        // term: self.term(),
+        // cb: callback,
+        // propose_time: Some(now),
+        // must_pass_epoch_check: false,
+        // };
+        //
+        // self.post_propose(poll_ctx, p);
+        // }
+        // }
+    }
+
+    // Returns a boolean to indicate whether the `read` is proposed or not.
+    // For these cases it won't be proposed:
+    // 1. The region is in merging or splitting;
+    // 2. The message is stale and dropped by the Raft group internally;
+    // 3. There is already a read request proposed in the current lease;
+    pub fn read_index<T: Transport>(
```

Comment:
How about inlining the function giving it actually has only 5 lines.

Reply:
The caller entry is too long to look good when it has repeated more than once. Also it has a TODO which will add more code into it later.  I prefer the current way. 
 Ok(RequestPolicy::ReadIndex) => {
                    self.fsm.peer_mut().read_index(self.store_ctx, req, ch);
                }
                vs.
                // TODO: add pre_read_index to handle splitting or merging
                 if self.fsm.peer_mut().is_leader() {
                      self.fsm.peer_mut().read_index_leader(poll_ctx, req, ch);
                  } else {
                      self.fsm.peer_mut().read_index_follower(poll_ctx, req, ch);
                  }

Reply:
but still it will have more code here. I think it's better to have a short and clear call in the switch.

Reply:
It won't. those code should be put into `read_index_leader` instead of the switch.

Reply:
As discussed, put the read_index to mod.rs.